python/semantic_kernel/utils/tracing/decorators.py

```diff
@@ -0,0 +1,233 @@
+# Copyright (c) Microsoft. All rights reserved.
+#
+# Code to trace model activities with the OTel semantic conventions.
+# This code contains experimental features and may change in the future.
+# To enable these features, set one of the following senvironment variables to true:
+#    SEMANTICKERNEL_EXPERIMENTAL_GENAI_ENABLE_OTEL_DIAGNOSTICS
+#    SEMANTICKERNEL_EXPERIMENTAL_GENAI_ENABLE_OTEL_DIAGNOSTICS_SENSITIVE
+
+import functools
+import json
+import os
+from collections.abc import Callable
+from typing import Any
+
+from opentelemetry import trace
+from opentelemetry.trace import Span, StatusCode
+
+from semantic_kernel.connectors.ai.prompt_execution_settings import PromptExecutionSettings
+from semantic_kernel.contents.chat_history import ChatHistory
+from semantic_kernel.contents.chat_message_content import ITEM_TYPES, ChatMessageContent
+from semantic_kernel.contents.function_call_content import FunctionCallContent
+from semantic_kernel.utils.tracing.const import (
+    COMPLETION_EVENT,
+    COMPLETION_EVENT_COMPLETION,
+    COMPLETION_TOKEN,
+    ERROR_TYPE,
+    FINISH_REASON,
+    MAX_TOKEN,
+    MODEL,
+    OPERATION,
+    PROMPT_EVENT,
+    PROMPT_EVENT_PROMPT,
+    PROMPT_TOKEN,
+    RESPONSE_ID,
+    SYSTEM,
+    TEMPERATURE,
+    TOP_P,
+)
+
+OTEL_ENABLED_ENV_VAR = "SEMANTICKERNEL_EXPERIMENTAL_GENAI_ENABLE_OTEL_DIAGNOSTICS"
+OTEL_SENSITIVE_ENABLED_ENV_VAR = "SEMANTICKERNEL_EXPERIMENTAL_GENAI_ENABLE_OTEL_DIAGNOSTICS_SENSITIVE"
+
+
+_enable_diagnostics = os.getenv(OTEL_ENABLED_ENV_VAR, "false").lower() in ("true", "1", "t")
+_enable_sensitive_events = os.getenv(OTEL_SENSITIVE_ENABLED_ENV_VAR, "false").lower() in ("true", "1", "t")
+
+# Creates a tracer from the global tracer provider
+tracer = trace.get_tracer(__name__)
+
+
+def are_model_diagnostics_enabled() -> bool:
+    """Check if model diagnostics are enabled.
+
+    Model diagnostics are enabled if either EnableModelDiagnostics or EnableSensitiveEvents is set.
+    """
+    return _enable_diagnostics or _enable_sensitive_events
+
+
+def are_sensitive_events_enabled() -> bool:
+    """Check if sensitive events are enabled.
+
+    Sensitive events are enabled if EnableSensitiveEvents is set.
+    """
+    return _enable_sensitive_events
+
+
+def trace_chat_completion(model_provider: str) -> Callable:
+    """Decorator to trace chat completion activities."""
+
+    def inner_trace_chat_completion(completion_func: Callable) -> Callable:
+        @functools.wraps(completion_func)
+        async def wrapper_decorator(*args: Any, **kwargs: Any) -> list[ChatMessageContent]:
+            chat_history: ChatHistory = kwargs["chat_history"]
+            settings: PromptExecutionSettings = kwargs["settings"]
+
+            if hasattr(settings, "ai_model_id") and settings.ai_model_id:
```

Comment:
why not use `getattr` with default None?

Reply:
I'm pretty new to Python...  Won't that cause an exception if ai_model_id does not exist?

Reply:
Ah, just read that it will throw only if no default value is passed.  Simplified.